tinygrad/codegen/kernel.py

```diff
@@ -78,16 +78,14 @@ def __init__(self, *ast:LazyOp, opts:Optional[CompilerOptions]=None):
     self.lazyops = flatten([op.lazyops for op in self.ast])
 
     # there's only allowed to be one reduceop
-    reduceops = [x for x in self.lazyops if x.op in ReduceOps]
-    assert len(dedup(reduceops)) <= 1, "max one reduce op in an ast"
-    self.reduceop = reduceops[0] if reduceops else None
+    self.reduceops = dedup(flatten([reversed([x for x in out.lazyops if x.op in ReduceOps]) for out in self.ast]))
```

Comment:
what ASTs do you plan on passing to the linearizer? how will you handle shapetracker pushing and reduceops?

Reply:
I pass all the asts to the linearizer, b/c we need all the outputs to get stored

tbh I didn't really need to change anything with the shapetracker; it only makes sense to fuse reducops from the same shape

the reduceops will get rendered sequentially this introduces the problem of:
1) loading back local reductions which gets solved [here](https://github.com/tinygrad/tinygrad/pull/4409/files#r1590002200)
2) scope, which with one reduce gets solved by `UOpGraph.fix_loop_scope` *except* that with multiple reduces `get_recursive_parents` will propagate all the way up to the local store of any prior reduceops, so we need to [cut off the recursion at barriers](https://github.com/tinygrad/tinygrad/pull/4409/files#r1590002961) (the parents of barriers aren't dependencies in the same sense as loops or alus)

I think we could handle different shape transitions, the reducops keep a notion of which axis they're reducing so there's enough information to manage everything in the linearizer alone

but same shape seemed to be the common case